docs/signature/README.md

```diff
@@ -0,0 +1,228 @@
+# Notary V2 Signature Specification
+
+This section defines the signature file, which is in JSON format with no whitespaces. Its JSON schema is available at [schema.json](schema.json).
```

Comment:
The formatting is the source of much conversation (not sure it's debate as nobody likes the JSON format).
In this case, we're trying to be explicit for how the signature is generated. But, we need to be more explicit. There's some reference info here: https://latacora.micro.blog/2019/07/24/how-not-to.html

Reply:
Yeah, the rules in that blog post are roughly what I was suggesting ("Serialize however you want.").  The existing proposal reads like an enforced serialization (the "Canonicalization" approach under the "How *not* to sign a JSON object, if you can help it" heading).

Reply:
Ok, that's good feedback if that's how you read it.

Reply:
I don't agree with that latacora blogpost for anything more than signing a single-keyed string JSON object (like most JWT payloads are anyway). I also don't think this'll hold for situations like "load, append a signature, re-serialize".

Signing an arbitrary string of bytes, like the blogpost suggests, is essentially expecting that we keep *this* particular serialization around (or a one to one mapping between verified and serialized to de-serialized objects). Again, this is not an issue if you're expecting a signed-ephemeral token, but it will become increasingly messy when we are talking about long-lived documents that may need multiple signatures.

Reply:
I'll set Shiwei, Cormack and Derek comment on the signature encoding. But, I think we are talking about signing a single verification object payload.
For multiple signatures, we don't intend to append to an existing document. Each signature object is immutable to assure the digests never change of the artifact they sign, or the verification object itself. In any larger, geo-dispersed model, attempting to update something is fraught with conflict resolution and constant chasing of current state. Rather, we're thinking a signature API will provide a paged collection of signature objects for a given artifact.

Reply:
> I also don't think this'll hold for situations like "load, append a signature, re-serialize".

What would be the need to re-serialize if we're not mutating the manifest itself?  As long as we retain the bytes that make up the original manifest, we can append new signatures by signing those same bytes.

> Signing an arbitrary string of bytes, like the blogpost suggests, is essentially expecting that we keep *this* particular serialization around

Registries already need to preserve the particular serialization of a manifest today, as the image digest depends on the bytes that make up the manifest.  Changing (re-serializing) the manifest would change the digest.

Reply:
I believe what Santiago means is that JSON serialization of data can be notoriously nondeterministic. Different parsers from different languages (or even different libraries) may serialize the same JSON object differently to disk. By signing a canonical string representation of a JSON object, we are able to consistently sign and verify, especially when passing the object around to be signed by different developers (who cannot safely be assumed to be using the same tools).

Reply:
> Different parsers from different languages (or even different libraries) may serialize the same JSON object differently to disk.

Yep, this is the challenge that the blog post talks about.

> By signing a canonical string representation of a JSON object, we are able to consistently sign and verify, especially when passing the object around to be signed by different developers (who cannot safely be assumed to be using the same tools).

Canonicalization has its own problems though; namely: you need a serializer that knows how to produce the canonical form rather than using a more widely-tested, widely-available serializer.

Instead of using a canonical string representation, if you preserve the original bytes and use that for signing and verifying you get the benefits of canonicalization without the downsides.  Implementations of the Docker registry (both client and server) as well as the OCI distribution spec (again both client and server) already need to have this behavior for correct digest calculation.

Reply:
Provided we can assume a standard, deterministic JSON serializer, we do not need canonicalization, but this just needs to be clearly documented.

Reply:
> Provided we can assume a standard, deterministic JSON serializer, we do not need canonicalization, but this just needs to be clearly documented.

We don't need a standard, deterministic serializer as long as we preserve the original serialization of the object.  For example, if I build an image and it gets serialized by by builder program, the registry will preserve the manifest bytes exactly as uploaded.  As long as we keep the original bytes around we should never need to re-serialize the object.  (We'll still need to de-serialize so we can access the data in the object, but that doesn't require re-serializing it again.)

Reply:
> We don't need a standard, deterministic serializer as long as we preserve the original serialization of the object. 
> ...
> As long as we keep the original bytes around we should never need to re-serialize the object.

This is exactly what I mean. Instead of having a deterministic serialization mechanism you're opting for keeping tabs on what is the blessed/signed payload and where it lives on disk (so you can use *that* instead when you want to sign it). Of course if the goal is to argue "once serialized, never re-serialized for no reason ever without resigning everything" because it makes a serialization bug less likely --- but more state-keeping code everywhere, and without introspection unless you attach the original payload everywhere --- then all I want is to put all the arguments forward.

 That aside, my very personal take is that I don't see how tiptoeing around the fact that "you know for *sure* that you can't reach  that same string twice for a *standardized object* is a win" in any shape or form.

Reply:
Right, not to mention that you're probably going to need to parse JSON anyway just to figure out where the signed message is and how long it is. To me, not at least [permitting canonicalization](https://github.com/theupdateframework/specification/pull/102) seems unnecessarily restrictive because it limits the JSON libraries used in practice, but we can decide this thorny issue later.

Reply:
Trishank, i was trying to reach you on slack to see how you solved this for CNAB?

Reply:
> Trishank, i was trying to reach you on slack to see how you solved this for CNAB?

Steve, IIRC we use Docker's [canonicaljson](https://github.com/docker/go/tree/master/canonical/json) @radu-matei

Reply:
That is correct - for Go projects in the CNAB ecosystem we use [`github.com/docker/go/canonical/json`](https://github.com/docker/go/tree/master/canonical/json), and other canonical JSON implementations for other languages.

I agree with the pain around JSON canonicalization (and JSON in general), but I also want to make sure this approach doesn't make it difficult to implement the libraries in other programming languages (for both signing and verifying).

Reply:
Please see jwt proposal: #3 
Resolving this in lieu of ^

Reply:
> In this PR, the signature is a JWT variant. Therefore, the signature file is no longer a JSON file.

@shizhMSFT I don't get it: how is JWT not JSON? Also, the PR seems to ignore canonicalization altogether for now.

Reply:
@trishankatdatadog, can we move this conversation to PR #3?